File: internal/checks/ssrf.go
package checks

import (
	"context"
	"fmt"
	"net/http"
	"net/url"
	"strings"

	"github.com/londonmax12/hyperz/internal/httpclient"
	"github.com/londonmax12/hyperz/internal/page"
	"github.com/londonmax12/hyperz/internal/scope"
)

// SSRF probes whether a target application will fetch an attacker-controlled
// URL via a URL parameter (e.g., image proxy, webhook, fetch endpoint). For
// each candidate Sink the check issues a no-follow request with the param
// set to a canary on a reserved (.example) domain; a response that contains
// certain error signatures (connection refused, name resolution failure,
// timeout, etc.) suggests the server attempted to reach the canary host.
//
// Candidates are chosen to bound probe volume:
//   - sinks with names suggesting URL acceptance (url, uri, fetch, proxy,
//     endpoint, webhook, etc.) are always probed.
//   - additional "generic" URL parameter names are only probed on URLs with
//     path keywords hinting at proxy/webhook functionality.
//
// This is an active (LevelDefault) check; it only runs when the user opts
// into a `default` or `aggressive` scan. Per-host rate limiting in the
// HTTP client governs probe pacing.
type SSRF struct{}

func (SSRF) Name() string { return "ssrf" }

func (SSRF) Level() Level { return LevelDefault }

const (
	// ssrfCanary uses RFC 2606 .example so the host is guaranteed
	// unregistered. The path marker makes the probe easy to spot in target
	// access logs.
	ssrfCanary  = "http://internal.example:8080/hyperz-probe"
	ssrfBodyCap = 8 << 10
)

// ssrfErrorPatterns are response content markers that indicate the server
// attempted to fetch the canary URL and encountered a network/DNS error.
// These signatures come from common libraries and frameworks across
// different languages (Python requests, Go http, Node.js, PHP curl, etc.).
var ssrfErrorPatterns = []string{
	// DNS failures
	"getaddrinfo failed",
	"nodename nor servname provided",
	"name or service not known",
	"no address associated with hostname",
	"temporary failure in name resolution",
	"host not found",
	"cannot resolve host",
	"unknown host",
	// Connection refused
	"connection refused",
	"econnrefused",
	"connection reset by peer",
	"reset by peer",
	// Connection timeouts
	"connection timed out",
	"operation timed out",
	"dial tcp",
	"timeout",
	// Proxy/fetch library errors
	"failed to fetch",
	"httperror",
	"socket timeout",
	"unable to connect",
	"unreachable",
	// Python requests specific
	"connectionerror",
	"requests.exceptions",
	// Node.js specific
	"enotfound",
	"request to",
	// Java specific
	"unknownhostexception",
	"connectexception",
	// PHP specific
	"failed to open stream",
	"could not resolve host",
	// Ruby specific
	"connection refused -- connect",
	"getaddrinfo",
	// Generic fetch/request patterns
	"fetch error",
	"request failed",
	"failed to request",
}

// ssrfSpecificParamNames are parameter names strongly indicating URL-fetch
// functionality. These are always probed.
var ssrfSpecificParamNames = []string{
	"url",
	"uri",
	"endpoint",
	"target",
	"fetch",
	"proxy",
	"image_url",
	"image_uri",
	"webhook",
	"callback",
	"callback_url",
	"callback_uri",
	"return_url",
	"return_uri",
	"source",
	"source_url",
	"destination",
	"request_url",
}

// ssrfGenericParamNames are additional parameter names only probed when
// the page looks like it handles proxying/webhooks based on path keywords.
var ssrfGenericParamNames = []string{
	"q",
	"query",
	"link",
	"page",
	"resource",
	"data",
	"content",
	"http",
}

// ssrfPathKeywords are path substrings that flag a URL as likely handling
// URL fetches (proxy, image handling, webhook receivers, etc.).
var ssrfPathKeywords = []string{
	"proxy",
	"fetch",
	"image",
	"avatar",
	"webhook",
	"callback",
	"export",
	"report",
	"download",
	"preview",
	"screenshot",
}

func (c SSRF) Run(ctx context.Context, client *httpclient.Client, sc *scope.Scope, p page.Page) ([]Finding, error) {
	target := p.URL
	u, err := url.Parse(target)
	if err != nil || u.Scheme == "" || u.Host == "" {
		return nil, nil
	}
	if !sc.Allows(u) {
		return nil, nil
	}

	sweep := LevelFrom(ctx) >= LevelAggressive || looksProxyish(u.Path)
	candidates := ssrfSinks(p, u, sweep)

	var findings []Finding
	var firstErr error
	for _, sink := range candidates {
		if ctx.Err() != nil {
			break
		}
		if u2, err := url.Parse(sink.URL); err == nil && !sc.Allows(u2) {
			continue
		}
		f, err := c.probe(ctx, client, target, sink)
		if err != nil {
			Report(ctx, fmt.Errorf("probe param %q: %w", sink.Name, err))
			if firstErr == nil {
				firstErr = err
			}
			continue
		}
		if f != nil {
			findings = append(findings, *f)
		}
	}
	if firstErr != nil && len(findings) == 0 {
		return nil, firstErr
	}
	return findings, nil
}

// ssrfSinks returns the deduped, sorted set of Sinks to probe.
func ssrfSinks(p page.Page, u *url.URL, sweep bool) []Sink {
	type key struct {
		method string
		url    string
		loc    Loc
		name   string
	}
	seen := map[key]struct{}{}
	add := func(out []Sink, s Sink) []Sink {
		if s.Name == "" || s.URL == "" {
			return out
		}
		k := key{s.Method, s.URL, s.Loc, s.Name}
		if _, ok := seen[k]; ok {
			return out
		}
		seen[k] = struct{}{}
		return append(out, s)
	}

	var out []Sink
	for _, s := range SinksFor(p) {
		out = add(out, s)
	}

	// Always probe sinks with names that strongly suggest URL handling
	for _, name := range ssrfSpecificParamNames {
		out = add(out, Sink{
			Method: http.MethodGet,
			URL:    p.URL,
			Loc:    LocQuery,
			Name:   name,
		})
	}

	// Probe generic names only when the page looks like it fetches URLs
	if sweep {
		for _, name := range ssrfGenericParamNames {
			out = add(out, Sink{
				Method: http.MethodGet,
				URL:    p.URL,
				Loc:    LocQuery,
				Name:   name,
			})
		}
	}

	return out
}

func looksProxyish(path string) bool {
	p := strings.ToLower(path)
	for _, kw := range ssrfPathKeywords {
		if strings.Contains(p, kw) {
			return true
		}
	}
	return false
}

// probe issues one no-follow request with the canary URL overlaid onto
// sink.Name. Returns a finding when the response body contains error
// signatures indicating the server attempted to fetch the canary host.
func (c SSRF) probe(ctx context.Context, client *httpclient.Client, target string, sink Sink) (*Finding, error) {
	req, err := sink.MutateRequest(ctx, ssrfCanary)
	if err != nil {
		return nil, err
	}
	resp, err := client.DoNoFollow(ctx, req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, truncated, err := httpclient.ReadBodyCapped(resp, ssrfBodyCap)
	if err != nil {
		return nil, err
	}

	// Check for SSRF error markers in the response body
	matchedPattern := ssrfMatchesError(body)
	if matchedPattern == "" {
		return nil, nil
	}

	probeURL := req.URL.String()
	return &Finding{
		Check:    c.Name(),
		Target:   target,
		URL:      probeURL,
		Severity: SeverityHigh,
		Title:    fmt.Sprintf("Server-Side Request Forgery via %s ?%s=", sink.Loc, sink.Name),
		Detail: fmt.Sprintf(
			"Parameter %q (%s) accepts a URL that the server fetches. "+
				"Probe with %s triggered server-side request attempt; "+
				"response contains error signature %q indicating connection failure. "+
				"An attacker can craft URLs to probe internal network, bypass authentication, or attack internal services.",
			sink.Name, sink.Loc, ssrfCanary, matchedPattern),
		CWE:   "CWE-918",
		OWASP: "A10:2021 Server-Side Request Forgery (SSRF)",
		Remediation: "Validate and restrict the URL parameter to a strict allowlist of domains/hosts. " +
			"Disable access to private/internal IP ranges (10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 127.0.0.0/8, ::1). " +
			"Use a URL parsing library that properly validates scheme and host. Never fetch arbitrary user-supplied URLs.",
		Evidence: &Evidence{
			Method:     req.Method,
			RequestURL: probeURL,
			Status:     resp.StatusCode,
			Exchange:   RecordExchange(req, nil, false, resp, body, truncated),
		},
		DedupeKey: MakeKey(c.Name(), ScopeParam, target, "loc:"+string(sink.Loc), "param:"+sink.Name),
	}, nil
}

// ssrfMatchesError returns the first matched error pattern from the body,
// or empty string if no match found. Comparison is case-insensitive.
func ssrfMatchesError(body []byte) string {
	bodyLower := strings.ToLower(string(body))
	for _, pattern := range ssrfErrorPatterns {
		if strings.Contains(bodyLower, pattern) {
			return pattern
		}
	}
	return ""
}

File: internal/checks/ssrf.lua
-- ssrf: for each candidate sink the probe overlays an attacker-
-- controlled URL on a reserved .example host and looks for HTTP-
-- library error signatures in the response body. A match means the
-- target's fetch primitive attempted the canary URL.
--
-- Two arms ride together when an OOB listener is attached:
--   1. In-band: plant the canary URL, scan body for error markers.
--      Catches targets that leak errors but cannot reach the OOB
--      listener (firewalled outbound, air-gapped network).
--   2. OOB: plant the listener's canary URL, drain on callback. Fires
--      a separate Critical-severity finding because an observed
--      callback proves both fetch + egress.
--
-- Sink selection: the always-probed "specific" names (url, fetch,
-- webhook, ...) cover well-known URL-handling parameters; the
-- "generic" sweep names (q, link, page, ...) are folded in when the
-- URL path looks proxy-ish OR the operator opted into aggressive
-- scanning.

local check = {
  name        = "ssrf",
  level       = "default",
  scope       = "param",
  cwe         = "CWE-918",
  owasp       = "A10:2021 Server-Side Request Forgery (SSRF)",
  remediation = "Validate and restrict the URL parameter to a strict allowlist of domains/hosts. "
                .. "Disable access to private/internal IP ranges (10.0.0.0/8, 172.16.0.0/12, 192.168.0.0/16, 127.0.0.0/8, ::1). "
                .. "Use a URL parsing library that properly validates scheme and host. Never fetch arbitrary user-supplied URLs.",
}

local function probe_inband(ctx, target, sink)
  local canary = ctx.payloads.ssrf_canary()
  local req, mut_err = sink:mutate_request(canary)
  if mut_err then return nil, mut_err end
  local resp, do_err = ctx.client:do_no_follow(req)
  if do_err then return nil, do_err end
  local body, truncated, rerr = resp:read_body_capped(ctx.payloads.ssrf_body_cap())
  if rerr then return nil, rerr end
  local pattern = ctx.body.ssrf_matches_error(body)
  if pattern == "" then return nil end

  local probe_url = req:url()
  return {
    severity = ctx.severity.high,
    target   = target,
    url      = probe_url,
    title    = string.format("Server-Side Request Forgery via %s ?%s=", sink.loc, sink.name),
    detail   = string.format(
      'Parameter %q (%s) accepts a URL that the server fetches. '
        .. 'Probe with %s triggered server-side request attempt; '
        .. 'response contains error signature %q indicating connection failure. '
        .. 'An attacker can craft URLs to probe internal network, bypass authentication, or attack internal services.',
      sink.name, sink.loc, canary, pattern),
    evidence = ctx.evidence.from_exchange {
      request   = req,
      response  = resp,
      body      = body,
      truncated = truncated,
    },
    dedupe_parts = { "loc:" .. sink.loc, "param:" .. sink.name },
  }
end

local function probe_oob(ctx, target, sink)
  if not ctx.oob:attached() then return end
  local canary = ctx.oob:register {
    target = target,
    sink   = sink.name,
    loc    = sink.loc,
    method = sink.method,
  }
  if canary == nil then return end
  local req, mut_err = sink:mutate_request(canary.http_url)
  if mut_err then
    ctx:report(string.format("oob probe param %q: %s", sink.name, mut_err))
    return
  end
  local resp, do_err = ctx.client:do_no_follow(req)
  if do_err then
    ctx:report(string.format("oob probe param %q: %s", sink.name, do_err))
    return
  end
  -- Drain a small chunk so the connection returns to the pool cleanly.
  -- The listener-side hit is the signal, not the response body.
  resp:read_body_capped(1024)
end

function check.run(ctx)
  local u, perr = ctx.url.parse(ctx.page.url)
  if perr or not u or u.scheme == "" or u.host == "" then return nil end
  if not ctx.scope:allows(ctx.page.url) then return nil end

  local sweep = ctx:level_at_least("aggressive") or ctx.payloads.ssrf_looks_proxyish(u.path)

  local sweep_params = {}
  for _, name in ipairs(ctx.payloads.ssrf_specific_params()) do
    sweep_params[#sweep_params + 1] = name
  end
  if sweep then
    for _, name in ipairs(ctx.payloads.ssrf_generic_params()) do
      sweep_params[#sweep_params + 1] = name
    end
  end

  local sinks = ctx.sinks.for_page { sweep_params = sweep_params }
  if #sinks == 0 then return nil end

  local findings = {}
  local first_err
  for _, sink in ipairs(sinks) do
    if ctx.scope:allows(sink.url) then
      local f, err = probe_inband(ctx, ctx.page.url, sink)
      if err then
        ctx:report(string.format("probe param %q: %s", sink.name, err))
        if not first_err then first_err = err end
      elseif f then
        findings[#findings + 1] = f
      end
      probe_oob(ctx, ctx.page.url, sink)
    end
  end

  if first_err and #findings == 0 then return nil, first_err end
  return findings
end

-- Drain emits one Critical-severity finding per OOB registration that
-- observed at least one callback. Callback evidence beats reflected-
-- error evidence, so OOB-confirmed sits at Critical while in-band
-- tops out at High.
function check.drain(ctx)
  if not ctx.oob:attached() then return nil end
  local findings = {}
  for _, reg in ipairs(ctx.oob:registrations()) do
    local hits = ctx.oob:hits(reg.token)
    if #hits > 0 then
      local extra = reg.extra or {}
      local target = extra.target or ""
      local sink_name = extra.sink or ""
      local loc = extra.loc or ""
      local method = extra.method or ""
      local hit = hits[1]
      findings[#findings + 1] = {
        severity = ctx.severity.critical,
        target   = target,
        url      = target,
        title    = string.format("Server-Side Request Forgery (OOB-confirmed) via %s %s", loc, sink_name),
        detail   = string.format(
          'Parameter %q (%s) accepts a URL that the server fetches. '
            .. 'Probe with canary %s caused the target to issue a request that landed on the OOB '
            .. 'listener (method=%s, source=%s, user-agent=%q, %d hit(s)). '
            .. 'An attacker can craft URLs to probe internal network, bypass authentication, '
            .. 'or attack internal services.',
          sink_name, loc, reg.http_url,
          hit.method, hit.source_addr, hit.user_agent or "", #hits),
        evidence = ctx.evidence.build {
          method  = method,
          url     = target,
          snippet = string.format(
            "Canary URL: %s\nFirst hit: %s %s from %s at %s\nUser-Agent: %s\nTotal hits: %d\n",
            reg.http_url, hit.method, hit.path, hit.source_addr,
            os.date("!%Y-%m-%dT%H:%M:%SZ", hit.timestamp_unix),
            hit.user_agent or "", #hits),
        },
        dedupe_parts = { "loc:" .. loc, "param:" .. sink_name, "oob" },
      }
    end
  end
  return findings
end

return check
